Adding Non-Maximum Suppression (NMS) utility to fiftyone.utils.labels through perform_nms()

rohis06 opened the pull request:
## What changes are proposed in this pull request?

Adds Non-Maximum Suppression (NMS) utility to fiftyone.utils.labels through `perform_nms()`! :tada:

It can be loaded as follows:
```python
import fiftyone.utils.labels as fl

dataset = foz.load_zoo_dataset(
    "coco-2017",
    split="validation",
    dataset_name="detector-recipe",
)

predictions_view = dataset.take(10, seed=51)

fl.perform_nms(sample_collection=predictions_view, in_field="predictions", out_field="nms_predictions", conf_threshold=0.65, iou_threshold=0.4, progress=True)
```

## How is this patch tested? If it is not, please explain why.

It has been thoroughly tested, as seen in this [nms_test.pdf](https://github.com/voxel51/fiftyone/files/14607543/nms_test.pdf)

## Release Notes

### Is this a user-facing change that should be mentioned in the release notes?

<!--
Please fill in relevant options below with an "x", or by clicking the checkboxes
after submitting this pull request. Example:
-   [x] Selected option
-->

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

Addition of perform_nms() utility to fiftyone.utils.labels

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [x] Documentation: FiftyOne documentation changes
-   [ ] Other

brimoor commented on fiftyone/utils/labels.py:
```diff
@@ -673,3 +674,116 @@ def classifications_to_detections(
+        sample_collection: a
+            :class:`fiftyone.core.collections.SampleCollection`
+        in_field: the name of the :class:`fiftyone.core.labels.Detections` field
+        out_field: the name of the :class:`fiftyone.core.labels.Detections`
```
How about making this optional: if not specified, the input field is updated in-place?

rohis06 replied:
Sure, that sounds like a good idea to me.
I'll make it optional.

brimoor commented on fiftyone/utils/labels.py:
```diff
@@ -673,3 +674,116 @@ def classifications_to_detections(
+
+    with fou.ProgressBar(progress=progress) as pb:
+        for sample in pb(samples):
+            detections_data_json_str = sample[in_field].to_json()
```
Can you implement this by processing `sample[in_field].detections` as a list of `Detection` objects rather than serializing to JSON? We generally try to avoid converting to other formats if not necessary.

Note that you can call `sample[in_field].copy()` to get a copy of the detections so that the existing detections aren't modified in-place

rohis06 replied:
Oh, definitely! I will check out the `.copy()` method.

brimoor commented on fiftyone/utils/labels.py:
```diff
@@ -673,3 +674,116 @@ def classifications_to_detections(
+        # Compare with other detections for NMS
+        for d in detections:
+            if d["label"] == selected_detection["label"]:
+                iou = _calculate_iou(
```
Can you use the existing `fiftyone.utils.iou.compute_ious()` utility here to avoid re-implementing this calculation?

rohis06 replied:
Sure, I'll incorporate this. It's much better always to use an existing utility! 😄

brimoor reviewed (changes requested):
@rohis06 overall the method looks great, just have a few suggestions on implementation that would clean things up a bit if you're open to them! 😄

rohis06 commented:
> @rohis06 overall the method looks great, just have a few suggestions on implementation that would clean things up a bit if you're open to them! 😄

Thanks a lot for reviewing this, @brimoor. I really appreciate your feedback! 😄
I'm definitely open to suggestions, and I'll incorporate them to clean things up.

rohis06 commented:
> > @rohis06 overall the method looks great, just have a few suggestions on implementation that would clean things up a bit if you're open to them! 😄
> 
> Thanks a lot for reviewing this, @brimoor. I really appreciate your feedback! 😄 I'm definitely open to suggestions, and I'll incorporate them to clean things up.

@brimoor, I have addressed the review comments.
Could you let me know if this clears things up? 😄

brimoor reviewed (approved):
@rohis06 thanks for your contribution! 🎉

rohis06 commented:
> @rohis06 thanks for your contribution! 🎉

Anytime! 😄
